lib/migration.js

```diff
@@ -412,92 +397,162 @@ function mixinMigration(MySQL, mysql) {
+        } else {
+          needsToDrop = true;
         }
-      }
-      if (oldSettings.Null === 'NO') {
-        // Did not allow null and now does.
-        if (self.isNullable(newSettings)) {
-          return true;
+
+        if (needsToDrop) {
+          sql.push('DROP FOREIGN KEY ' + fk.fkName);
+          removedFks.push(fk); //keep track that we removed these
         }
-      }
+      });
 
-      if (oldSettings.Type.toUpperCase() !==
-        self.buildColumnType(newSettings).toUpperCase()) {
-        return true;
-      }
-      return false;
+      //update out list of existing keys by removing dropped keys
+      fks = actualFks.filter(function(k) {
+        return removedFks.indexOf(k) == -1;
+      });
     }
+    if (sql.length > 0) {
+      self.applySqlChanges(model, sql, function(err, result) {
+        if (err) cb(err);
+        else
+          cb(err, fks);
+      });
+    } else cb(null, fks);
+  };
 
-    function expectedColName(propName) {
-      return expectedColNameForModel(propName, m);
+  MySQL.prototype.alterTable = function(model, actualFields, actualIndexes, actualFks, done, checkOnly) {
+    //if this is using an old signature, then grab the correct callback and check boolean
+    if ('function' == typeof actualFks && typeof done !== 'function') {
+      checkOnly = done || false;
+      done = actualFks;
     }
+    var self = this;
+    var m = this.getModelDefinition(model);
 
-    function expectedColNameForModel(propName, modelToCheck) {
-      var mysql = modelToCheck.properties[propName].mysql;
-      if (typeof mysql === 'undefined') {
-        return propName;
-      }
-      var colName = mysql.columnName;
-      if (typeof colName === 'undefined') {
-        return propName;
+    var propNames = Object.keys(m.properties).filter(function(name) {
+      return !!m.properties[name];
+    });
+
+    var indexNames = m.settings.indexes ? Object.keys(m.settings.indexes).filter(function(name) {
+      return !!m.settings.indexes[name];
+    }) : [];
+
+    var applyPending = function(actions, cb, err, result) {
+      var action = actions.shift();
+      var pendingChanges = action && action.call(self, model, actualFields) || [];
+      if (pendingChanges.length > 0) {
+        self.applySqlChanges(model, pendingChanges, function(err, result) {
+          if (!err) {
+            applyPending(actions, cb, err, result);
+          } else {
+            cb(err, result);
+          }
+        });
+      } else if (actions.length) {
+        applyPending(actions, cb); // We still have actions to apply
+      } else {
+        cb(err, result); // All finished
       }
-      return colName;
-    }
-  };
+    };
 
-  MySQL.prototype.getAlterStatement = function(model, statements) {
-    return statements.length ?
-      'ALTER TABLE ' + this.tableEscaped(model) + ' ' + statements.join(',\n') :
-      '';
+    async.series([
+      function(cb) {
+        applyPending([self.getAddModifyColumns], cb);
```

Comment:
@jannyHou I disagree. `async.series` invokes functions serially (waiting for each preceding one to finish before starting next), hence it will execute `getAddModifyColumns` first and then `getDropColumns`.